src/mame/misc/xtom3d_piu10.cpp

```diff
@@ -0,0 +1,206 @@
+    U20 UTC 78L05 WK
+    U21, U22 HA17358
+
+    CN1 PCN96, connects to SPACE12_1
+    CN3, CN4 Unpopulated 4 pin headers
+
+    14.7456MHz XTAL near Micronas chips
+
+
+    CAT702 security data:
+    https://github.com/pumpitupdev/pumptools/blob/b809d8674925d3c6e6b25efa91dccc04a460a62f/src/main/sec/lockchip/lockchip-defs.h
+ */
+
+#include "emu.h"
+#include "xtom3d_piu10.h"
+
+
+DEFINE_DEVICE_TYPE(ISA16_PIU10, isa16_piu10, "isa16_piu10", "ISA16 PIU10 for MK-III")
+
+isa16_piu10::isa16_piu10(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, ISA16_PIU10, tag, owner, clock)
+	, device_isa16_card_interface(mconfig, *this)
+	, m_cat702(*this, "cat702")
+	, m_dac3350a(*this, "dac3350a")
+	, m_mas3507d(*this, "mas3507d")
+	, m_flash(*this, "flash_u8")
+{
+}
+
+void isa16_piu10::device_start()
+{
+	set_isa_device();
+
+	save_item(NAME(m_addr));
+	save_item(NAME(m_dest));
+	save_item(NAME(m_flash_unlock));
+	save_item(NAME(m_mp3_demand));
+	save_item(NAME(m_mp3_mpeg_frame_sync));
+	save_item(NAME(m_cat702_data));
+}
+
+void isa16_piu10::device_reset()
+{
+	m_addr = m_dest = 0;
+	m_flash_unlock = false;
+
+	m_mp3_demand = 1;
+	m_mp3_mpeg_frame_sync = 1;
+
+	m_cat702_data = 0;
+}
+
+void isa16_piu10::device_add_mconfig(machine_config &config)
+{
+	MACRONIX_29F1610MC_16BIT(config, m_flash);
+
+	CAT702_PIU(config, m_cat702, 0);
+	m_cat702->dataout_handler().set([this] (u16 data) {
+		m_cat702_data = data & 1;
+	});
+
+	DAC3350A(config, m_dac3350a);
+
+	MAS3507D(config, m_mas3507d);
+	m_mas3507d->mpeg_frame_sync_cb().set(*this, FUNC(isa16_piu10::mas3507d_mpeg_frame_sync));
+	m_mas3507d->demand_cb().set(*this, FUNC(isa16_piu10::mas3507d_demand));
+	m_mas3507d->add_route(0, m_dac3350a, 1.0, 0);
+	m_mas3507d->add_route(1, m_dac3350a, 1.0, 1);
+}
+
+void isa16_piu10::remap(int space_id, offs_t start, offs_t end)
+{
+	if (space_id == AS_IO)
+	{
+		m_isa->install16_device(0x02d0, 0x02df, read16sm_delegate(*this, FUNC(isa16_piu10::read)), write16sm_delegate(*this, FUNC(isa16_piu10::write)));
+	}
+}
+
+uint16_t isa16_piu10::read(offs_t offset)
+{
+	switch (offset)
+	{
+		case 0x5: // 2da
+			if (m_dest == 0x008)
+			{
+				uint16_t r = 0;
+
+				r |= (m_cat702_data & 1) << 5;
+
+				r |= m_mp3_mpeg_frame_sync << 2; // every 256th transition the game will set the internal playback position timestamp to counter*6.2687998 to sync with the MP3 decoder chip (counter*6.2687998/240 = (1152*counter)/44100)
+				r |= 1 << 1; // the MP3 data send function loops until this is 1 before sending MP3 data byte
+				r |= m_mp3_demand << 0; // if not set then MP3 data send function early returns
+
+				return r;
+			}
+			else if (m_dest == 0)
+			{
+				const uint32_t offs = m_addr;
+				if (m_flash_unlock)
+					m_addr++;
+				return m_flash->read(offs);
+			}
+			break;
+	}
+
+	logerror("%s: unknown read %d %03x %06x %02x\n", machine().describe_context(), m_flash_unlock, m_dest, m_addr, offset);
```

Comment:
Remember to check `machine().side_effects_disabled()` before applying read side effects to avoid chaos with the debugger (incrementing address).  It’s also a good idea to check it before logging unmapped reads to avoid spam when memory windows are open.